OpenRA.Mods.D2k/Traits/Buildings/D2kBuilding.cs

```diff
@@ -57,12 +67,51 @@ public D2kBuilding(ActorInitializer init, D2kBuildingInfo info)
 		void INotifyCreated.Created(Actor self)
 		{
 			health = self.TraitOrDefault<IHealth>();
+			layer = self.World.WorldActor.TraitOrDefault<BuildableTerrainLayer>();
+			techTree = self.Owner.PlayerActor.TraitOrDefault<TechTree>();
 		}
 
 		protected override void AddedToWorld(Actor self)
 		{
 			base.AddedToWorld(self);
 
+			if (layer != null && (!info.ConcretePrerequisites.Any() || techTree == null || techTree.HasPrerequisites(info.ConcretePrerequisites)))
+			{
+				var map = self.World.Map;
+				var template = map.Rules.TileSet.Templates[info.ConcreteTemplate];
+				if (template.PickAny)
+				{
+					// Fill the footprint with random variants
+					foreach (var c in info.Tiles(self.Location))
+					{
+						if (!map.Contains(c) || map.CustomTerrain[c] != byte.MaxValue)
+							continue;
+
+						// Don't place under other buildings
+						if (self.World.ActorMap.GetActorsAt(c).Any(a => a != self && a.TraitOrDefault<Building>() != null))
```

Comment:
Depending on what we do with #18824, this could use `self.World.ActorMapIsBuildingBlocked` (after moving this block above the `base.AddedToWorld`)?